### PR TITLE
Added Support for 3D Reconstruction

### PR DESCRIPTION
**Pull request by chaithyagr**
This branch generalizes the flow for 3D reconstruction.
I have added an example python script too.
I have seen that sometimes the cost increases if we don't have sufficient iterations in the calculation of Lipschitz Constant. Any ideas on how to fix this?

### CODE REVIEW

File: mri/reconstruct/utils.py [hunk 1]
```diff
@@ -222,3 +247,23 @@ def generate_operators(data, wavelet_name, samples, mu=1e-06, nb_scales=4,
             plot_output=None)
 
     return gradient_op, linear_op, prox_op, cost_op
+
+
+def imshow3D(volume, display=True):
```
Comment (zaccharieramzi) [inline mri/reconstruct/utils.py]: I think this doesn't really belong in reconstruct. I think we could put it in a more general utils at the root level.

File: mri/reconstruct/utils.py [hunk 2]
```diff
@@ -222,3 +247,23 @@ def generate_operators(data, wavelet_name, samples, mu=1e-06, nb_scales=4,
...
+    """ Display an 3D volume on the axes, press "p" or "n" to navigate across
+    the slices.
+
+    Parameters
```
Comment (zaccharieramzi) [inline mri/reconstruct/utils.py]: Document all the parameters.

File: mri/reconstruct/utils.py [hunk 3]
```diff
@@ -14,9 +14,34 @@
 
 # System import
 import numpy as np
+import matplotlib.pyplot as plt
 import warnings
 
 
+class IndexTracker(object):
```
Comment (zaccharieramzi) [inline mri/reconstruct/utils.py]: When you move that class, don't forget to document it a bit.

File: mri/reconstruct/fourier.py [hunk 4]
```diff
@@ -209,3 +215,221 @@ def adj_op(self, x):
         """
         self.plan.f = x
         return np.copy(self.plan.adjoint()) / np.sqrt(self.plan.M)
+
+
+class Singleton:
```
Comment (zaccharieramzi) [inline mri/reconstruct/fourier.py]: can you document that class? I feel that this a bit like a counter, am I right?

File: mri/reconstruct/fourier.py [hunk 5]
```diff
@@ -209,3 +215,221 @@ def adj_op(self, x):
...
+        Parameters
+        ----------
+        samples: np.ndarray
+            the mask samples in the Fourier domain.
+        shape: tuple of int
+            shape of the image (necessarly a square/cubic matrix).
+        platform: string, 'cpu', 'mcpu' or 'gpu'
+            string indicating which hardware platform will be used to
+            compute the NUFFT
+        Kd: int or tuple
```
Comment (zaccharieramzi) [inline mri/reconstruct/fourier.py]: Are these names universal? If not, I think we would benefit from having longer variables names.

File: mri/reconstruct/fourier.py (hunk 4 above)
Comment (chaithyagr) [reply mri/reconstruct/fourier.py]: Yes, it tells about the number of instances.

File: mri/test_local/test_wavelet_adjoint.py [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+            f = (numpy.random.randn(*f_p.shape) +
+                 1j * numpy.random.randn(*f_p.shape))
+            I_p = wavelet_op_adj.adj_op(f)
+            x_d = numpy.dot(Img.flatten(), numpy.conj(I_p).flatten())
+            x_ad = numpy.dot(f_p.flatten(), numpy.conj(f).flatten())
+            mismatch = (1. - numpy.mean(
+                numpy.isclose(x_d, x_ad,
+                              rtol=1e-6)))
+            print("      mismatch = ", mismatch)
+            self.assertTrue(numpy.isclose(x_d, x_ad, rtol=1e-6))
```
[the remark this answers is not in the document]
Comment (chaithyagr) [reply mri/test_local/test_wavelet_adjoint.py]: Yes, I am making these changes. Earlier tests have asserts before prints, it causes issues in debug.

File: mri/reconstruct/fourier.py (hunk 4 above)
Comment (zaccharieramzi) [reply mri/reconstruct/fourier.py]: The number of instances of what? Could you use `Counter` from `collections`?

**Comment by chaithyagr** [comment]
@LElgueddari , in the recent commit, I have pulled in changes requested by @zaccharieramzi . I think you can review now when you are free.

**Comment by LElgueddari** [comment]
Could you rebase your work?
Thank you

**Comment by chaithyagr** [comment]
Yes, will rebase and split up this PR. That makes more sense than reviewing everything. Closing this PR now.

**Pull request closed without merging**